src/impl/tlstransport.cpp

```diff
@@ -296,7 +296,224 @@ int TlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* ms
+	mbedtls_ssl_config_init(&mConf);
+	mbedtls_ctr_drbg_set_prediction_resistance(&mDrbg, MBEDTLS_CTR_DRBG_PR_ON);
+
+	try {
+		mbedtls::check(mbedtls_ctr_drbg_seed(&mDrbg, mbedtls_entropy_func, &mEntropy, NULL, 0));
+
+		mbedtls::check(mbedtls_ssl_config_defaults(
+		    &mConf, mIsClient ? MBEDTLS_SSL_IS_CLIENT : MBEDTLS_SSL_IS_SERVER,
+		    MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT));
+
+		mbedtls_ssl_conf_authmode(&mConf, MBEDTLS_SSL_VERIFY_OPTIONAL);
+		mbedtls_ssl_conf_rng(&mConf, mbedtls_ctr_drbg_random, &mDrbg);
+
+		if (certificate != nullptr) {
+			auto [crt, pk] = certificate->credentials();
+			mbedtls::check(mbedtls_ssl_conf_own_cert(&mConf, crt.get(), pk.get()));
+		}
+
+		mbedtls::check(mbedtls_ssl_setup(&mSsl, &mConf));
+		mbedtls_ssl_set_bio(&mSsl, static_cast<void *>(this), WriteCallback, ReadCallback, NULL);
+	} catch (...) {
+		mbedtls_entropy_free(&mEntropy);
+		mbedtls_ctr_drbg_free(&mDrbg);
+		mbedtls_ssl_free(&mSsl);
+		mbedtls_ssl_config_free(&mConf);
+		throw;
+	}
+}
+
+TlsTransport::~TlsTransport() {}
+
+void TlsTransport::start() {
+	PLOG_DEBUG << "Starting TLS transport";
+	registerIncoming();
+	changeState(State::Connecting);
+	enqueueRecv(); // to initiate the handshake
+}
+
+void TlsTransport::stop() {
+	PLOG_DEBUG << "Stopping TLS transport";
+	unregisterIncoming();
+	mIncomingQueue.stop();
+	enqueueRecv();
+}
+
+bool TlsTransport::send(message_ptr message) {
+	if (state() != State::Connected)
+		throw std::runtime_error("TLS is not open");
+
+	if (!message || message->size() == 0)
+		return outgoing(message); // pass through
+
+	PLOG_VERBOSE << "Send size=" << message->size();
+
+	mbedtls::check(mbedtls_ssl_write(
+	    &mSsl, reinterpret_cast<const unsigned char *>(message->data()), int(message->size())));
+
+	return true;
+}
+
+void TlsTransport::incoming(message_ptr message) {
+	if (!message) {
+		mIncomingQueue.stop();
+		enqueueRecv();
+		return;
+	}
+
+	PLOG_VERBOSE << "Incoming size=" << message->size();
+	mIncomingQueue.push(message);
+	enqueueRecv();
+}
+
+bool TlsTransport::outgoing(message_ptr message) { return Transport::outgoing(std::move(message)); }
+
+void TlsTransport::postHandshake() {
+	// Dummy
+}
+
+void TlsTransport::doRecv() {
+	std::lock_guard lock(mRecvMutex);
+	--mPendingRecvCount;
+
+	if (state() != State::Connecting && state() != State::Connected)
+		return;
+
+	try {
+		const size_t bufferSize = 4096;
+		char buffer[bufferSize];
+
+		// Handle handshake if connecting
+		if (state() == State::Connecting) {
+			auto handshakeStatus = mbedtls_ssl_handshake(&mSsl);
+			switch (handshakeStatus) {
+			case 0:
+				break;
+			case MBEDTLS_ERR_SSL_WANT_WRITE:
+			case MBEDTLS_ERR_SSL_WANT_READ:
+			case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+			case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
+				auto timeout = milliseconds(100);
+				ThreadPool::Instance().schedule(timeout, [weak_this = weak_from_this()]() {
+					if (auto locked = weak_this.lock())
+						locked->doRecv();
+				});
```

Comment:
This should be removed as the connection is reliable and there are no retransmissions here.

Reply:
I believe we still need this @paullouisageneau when running locally I would occasionally see handshake failures without it.  

```
   * \return         #MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS if a cryptographic
   *                 operation is in progress (see mbedtls_ecp_set_max_ops()) -
   *                 in this case you must call this function again to complete
   *                 the handshake when you're done attending other tasks.
```

Reply:
Calling again after an arbitrary timeout is kind of a hack and could indicate a return code is not handled as expected currently. Do you still see handshake failures after fixing `WriteCallback` not to incorrectly return `MBEDTLS_ERR_SSL_WANT_READ`? This could have caused the reception to wait for input and stall instead of reading available incoming data.

If I understand the doc correctly, the `X_IN_PROGRESS` functions require the user to call again as soon as possible, so there should be no wait in those cases (same for DTLS). Also, it sounds like `MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS` may only be returned if `mbedtls_ecp_set_max_ops()` was called to limit the number of operations before interrupting as the default is unlimited, so I guess it's never returned here.

Reply:
I still failures after the `WriteCallback` changed

I updated the code to call again immediately on the suggest return codes (instead of timing out)